cellxgene_schema_cli/tests/test_validate.py

```diff
@@ -0,0 +1,408 @@
+        self.validator.adata.X = numpy.zeros(self.validator.adata.X.shape)
+        self.validator._validate_sparsity()
+        self.assertTrue(self.validator.warnings)
+
+        self.validator.warnings = []
+        self.validator.adata.X = sparse.csc_matrix(self.validator.adata.X)
+        self.validator._validate_sparsity()
+        self.assertTrue(self.validator.warnings)
+
+        # Correct sparsity
+        self.validator.warnings = []
+        self.validator.adata.X = sparse.csr_matrix(self.validator.adata.X)
+        self.validator._validate_sparsity()
+        self.assertFalse(self.validator.warnings)
+
+
+class TestObsmValidation(unittest.TestCase):
+
+    def setUp(self):
+        self.validator = validate.Validator()
+        self.validator.adata = examples.adata.copy()
+        self.schema_def = validate._get_schema_definition(examples.SCHEMA_VERSION)
+        self.good_uns = examples.good_uns
+
+    def test_validate_good_obsm(self):
+        self.validator._validate_embedding_dict()
+        self.assertFalse(self.validator.errors)
+
+    def test_validate_bad_obsm(self):
+
+        # Wrong dimension, delete one column
+        key = list(self.validator.adata.obsm.keys())[0]
+        self.validator.adata.obsm[key] = numpy.delete(self.validator.adata.obsm[key], 0, axis=1)
+
+        self.validator._validate_embedding_dict()
+        self.assertTrue(self.validator.errors)
+
+    def test_validate_bad_obsm_2(self):
+
+        # Wrong type
+        key = list(self.validator.adata.obsm.keys())[0]
+        self.validator.adata.obsm[key] = pd.DataFrame(self.validator.adata.obsm[key],
+                                                      index = self.validator.adata.obs_names)
+
+        self.validator._validate_embedding_dict()
+        self.assertTrue(self.validator.errors)
+
+
+class TestUnsValidation(unittest.TestCase):
+
+    def setUp(self):
+        self.validator = validate.Validator()
+        self.validator.adata = examples.adata
+        self.schema_def = validate._get_schema_definition(examples.SCHEMA_VERSION)
+        self.good_uns = examples.good_uns
+        self.bad_uns = examples.bad_uns
+
+    def test_validate_good_uns(self):
+        self.validator._validate_dict(
+            self.good_uns, "uns", self.schema_def["components"]["uns"]
+        )
+        self.assertFalse(self.validator.errors)
+
+    def test_validate_bad_uns(self):
+
+        # Do one key at a time but skip schema_version as that's not checked here
+        for key, value in self.bad_uns.items():
+
+            if key == "schema_version":
+                continue
+
+            current_dict = self.good_uns.copy()
+            current_dict[key] = value
+
+            self.validator.errors = []
+            self.validator._validate_dict(
+                current_dict, "uns", self.schema_def["components"]["uns"]
+            )
+            self.assertTrue(self.validator.errors)
+
+
+class TestColumnValidation(unittest.TestCase):
```

Comment:
suggest `TestObsAndVarColumnValidation` for quick identification of the scope of this class.